eventlog-counters/src/GHC/Eventlog/Counters.hs

```diff
@@ -0,0 +1,249 @@
+
+    cntGCStart    <- newTVar Nothing
+    cntMaxGCTime  <- newTVar 0
+    cntGcCount0   <- newTVar 0
+    cntGcCountS   <- newTVar 0
+
+    cntThreads    <- newTVar Map.empty
+
+    return Counters {..}
+
+count :: Counters -> Event -> STM ()
+count cnt (Event t s c) = do
+    -- Count events
+    modifyTVar' (cntEvents cnt) (+ 1)
+
+    -- Update time
+    t' <- readTVar (cntTime cnt)
+    writeTVar (cntTime cnt) (max t t')
+
+    -- update seen capabilities
+    for_ c $ \c' -> modifyTVar' (cntCaps cnt) (IS.insert c')
+
+    case s of
+        -- We shouldn't see this event anymore
+        EventBlock {} -> return ()
+
+        -- real time
+        WallClockTime { sec, nsec} ->  writeTVar (cntWallClock cnt) $ Just WallClockInfo
+            { wciTime = t
+            , wciSec  = sec
+            , wciNSec = nsec
+            }
+
+        -- Thread events
+        WakeupThread {..}  -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadQueued)
+        CreateThread {..}  -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadQueued)
+        RunThread {..}     -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadRunning)
+        MigrateThread {..} -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadUnknown)
+        ThreadLabel {..}   -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadUnknown)
+        StopThread {..}    -> case status of
+          ThreadFinished -> modifyTVar' (cntThreads cnt) (Map.delete thread)
+          ForeignCall    -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadQueued)
+          HeapOverflow   -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadQueued)
+          StackOverflow  -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadQueued)
+          _              -> modifyTVar' (cntThreads cnt) (Map.insertWith (<>) thread ThreadStopped)
+
+        ThreadRunnable {} -> return ()
+
+        -- GC events
+        StartGC      -> writeTVar (cntGCStart cnt) (Just t)
+        GCWork       -> return ()
+        GCIdle       -> return ()
+        GCDone       -> return ()
+        GlobalSyncGC -> return ()
+        EndGC        ->  do
+            mt <- readTVar (cntGCStart cnt)
+            case mt of
+                Nothing    -> return () -- shouldn't happen, but may.
+                Just start -> modifyTVar' (cntMaxGCTime cnt) (max (t - start))
+
+        -- GC requests
+        RequestSeqGC -> return ()
+        RequestParGC -> return ()
+
+        -- stats
+        GCStatsGHC {..}                  ->  do
+            if gen == 0 then modifyTVar' (cntGcCount0 cnt) (+ 1) else modifyTVar' (cntGcCountS cnt) (+ 1)
+
+        HeapAllocated { allocBytes = x } -> writeTVar (cntHeapAlloc cnt) x
+        HeapSize { sizeBytes = x }       -> writeTVar (cntHeapSize cnt) x
+        BlocksSize { blocksSize = x }    -> writeTVar (cntBlocksSize cnt) x
+        MemReturn {}                     -> return () -- TODO
```

Comment:
Still these TODO comments. Do you plan to add these events?

Reply:
I'll add `MemReturn` but not `HeapInfoGHC` for now. it seems we should have the information latter provides already from other events.

Reply:
Done